server/src/test/java/org/elasticsearch/env/NodeEnvironmentTests.java

```diff
@@ -470,6 +471,72 @@ public void testExistingTempFiles() throws IOException {
         }
     }
 
+    public void testFailOnDataForNonDataNode() throws IOException {
+        Settings settings = buildEnvSettings(Settings.EMPTY);
+        final NodeEnvironment env = newNodeEnvironment(settings);
+
+        Index badIndex = new Index("bad", "badUUID");
+        for (Path path : env.indexPaths(badIndex)) {
+            Files.createDirectories(path.resolve("0"));
+        }
+
+        env.close();
+
+        try {
+            Settings noDataSettings = Settings.builder()
+                .put(settings)
+                .put(Node.NODE_DATA_SETTING.getKey(), false).build();
+            newNodeEnvironment(noDataSettings).close();
+            fail("Must fail creating NodeEnvironment on a data path that has shard data if node.data=false");
+        } catch (IllegalArgumentException e) {
+            assertThat(e.getMessage(), containsString(env.indexPaths(badIndex)[0].resolve("0").getFileName().toString()));
+        }
+    }
+
+    public void testSucceedOnMetaOnlyForNonDataNode() throws IOException {
+        Settings settings = buildEnvSettings(Settings.EMPTY);
+        final NodeEnvironment env = newNodeEnvironment(settings);
+
+        Index goodIndex = new Index("good", "goodUUID");
+
+        for (Path path : env.indexPaths(goodIndex)) {
+            Files.createDirectories(path.resolve(MetaDataStateFormat.STATE_DIR_NAME));
+        }
+
+        env.close();
+
+        Settings noDataSettings = Settings.builder()
+            .put(settings)
+            .put(Node.NODE_DATA_SETTING.getKey(), false).build();
+
+        // test that this succeeds
+        newNodeEnvironment(noDataSettings).close();
+    }
+
+    public void testSucceedOnDataForDataNode() throws IOException {
```

Comment:
maybe merge this test into `testFailOnDataForNonDataNode` above, calling it `test EnsureNoShardData`